tuf/tuf.go

```diff
@@ -173,6 +173,132 @@ func (tr *Repo) RemoveBaseKeys(role string, keyIDs ...string) error {
+	}
+	_, ok := tr.Root.Signed.Roles[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, fmt.Errorf("TUF data for targets does not exist in root metadata")
+	}
+	// Traverse target metadata, down to delegation itself
+	// Get all public keys for the base role from TUF metadata
+	signedTargetData, ok := tr.Targets[data.CanonicalTargetsRole]
+	if !ok {
+		return data.DelegationRole{}, fmt.Errorf("TUF data for targets does not exist in targets metadata", name)
+	}
+
+	// Keep track of paths and path hash prefixes to validate as we traverse the delegations tree, targets implicitly has ""
+	whiteListedPaths := []string{""}
+	whiteListedPathHashes := []string{""}
+
+	// Start with top level roles in targets
+	delegationRoles := signedTargetData.Signed.Delegations.Roles
+	var foundRole *data.Role
+	for len(delegationRoles) > 0 {
+		delgRole := delegationRoles[0]
+		delegationRoles = delegationRoles[1:]
+
+		// If this role is delegated above or is our desired role, validate paths and traverse its child roles
+		if delgRole.Name == name || strings.HasPrefix(name, delgRole.Name+"/") {
+			if okPaths := validateDelegationPathPrefixes(whiteListedPaths, delgRole.Paths); !okPaths {
+				return data.DelegationRole{}, fmt.Errorf("Invalid paths specified for delegation %s", delgRole.Name)
+			}
+			whiteListedPaths = delgRole.Paths
+			if okPathHashPrefixes := validateDelegationPathPrefixes(whiteListedPathHashes, delgRole.PathHashPrefixes); !okPathHashPrefixes {
+				return data.DelegationRole{}, fmt.Errorf("Invalid path hash prefixes specified for delegation %s", delgRole.Name)
+			}
+
+			// If we found the role, we can exit the loop
+			if delgRole.Name == name {
+				foundRole = delgRole
+				break
+			}
+
+			// If this is a parent role, keep traversing
+			whiteListedPathHashes = delgRole.PathHashPrefixes
+			delegationRoles = append(delegationRoles, tr.Targets[delgRole.Name].Signed.Delegations.Roles...)
+		}
+	}
+
+	if foundRole == nil {
+		return data.DelegationRole{}, fmt.Errorf("could not find valid delegation for role %s", name)
+	}
+
+	pubKeys := make(map[string]data.PublicKey)
+	parentData, ok := tr.Targets[path.Dir(foundRole.Name)]
+	if !ok {
+		return data.DelegationRole{}, fmt.Errorf("delegation parent data for %s does not exist in targets metadata", foundRole.Name)
+	}
+	for _, keyID := range foundRole.KeyIDs {
+		pubKey, ok := parentData.Signed.Delegations.Keys[keyID]
+		if !ok {
+			return data.DelegationRole{}, fmt.Errorf("delegation %s does not exist in targets metadata", foundRole.Name)
+		}
+		pubKeys[keyID] = pubKey
+	}
+
+	return data.DelegationRole{
+		BaseRole: data.BaseRole{
+			Name:      name,
+			Keys:      pubKeys,
+			Threshold: foundRole.Threshold,
+		},
+		Paths:            foundRole.Paths,
+		PathHashPrefixes: foundRole.PathHashPrefixes,
+	}, nil
+}
+
+// Returns true if the delegationPaths are prefixed by parentPaths
+func validateDelegationPathPrefixes(parentPaths, delegationPaths []string) bool {
```

Comment:
:+1: Awesome, this will be super useful, thanks for fixing the path prefixing stuff too!